suite-sparse ^openblas ~shared threads=openmp: undefined reference to `omp_get_max_threads'

### Steps to reproduce the issue

Building `suite-sparse@5.13.0 ^openblas ~shared threads=openmp` fails using:
* `spack@develop` (b8735fd1e49006f5b48d516309168366298fe687 from `Tue Mar 28 11:13:26 2023 -0600`)
* Ubuntu 20.04, X86_64
* GCC 11.1.0

Spack environment: [suite-sparse.spack.yaml.txt](https://github.com/spack/spack/files/11099940/suite-sparse.spack.yaml.txt)

<details><summary>Concretization</summary><pre>
 -   6ptvibp  suite-sparse@5.13.0%gcc@11.1.0~cuda~graphblas+openmp+pic build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  l3pqtzc      ^gmp@6.2.1%gcc@11.1.0+cxx build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  vlza2rq          ^autoconf@2.69%gcc@11.1.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
[+]  tflyjjh          ^automake@1.16.5%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  yyuvbzt          ^libtool@2.4.7%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  t4wviga      ^m4@1.4.19%gcc@11.1.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
[+]  j2tlgt4          ^diffutils@3.9%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  5f66p4a              ^libiconv@1.17%gcc@11.1.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  b2tzsvo          ^libsigsegv@2.14%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  h32qvnm      ^metis@5.1.0%gcc@11.1.0~gdb~int64~ipo~real64+shared build_system=cmake build_type=RelWithDebInfo generator=make patches=4991da9,93a7903,b1225da arch=linux-ubuntu20.04-x86_64
[+]  viujxrb          ^cmake@3.25.2%gcc@11.1.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64
[+]  lwdu2k7              ^ncurses@6.3%gcc@11.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  uorrjkf                  ^pkgconf@1.8.0%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  chqnfsn              ^openssl@1.1.1t%gcc@11.1.0~docs~shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64
[+]  w233qs7                  ^ca-certificates-mozilla@2023-01-10%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  ac6p2ol          ^gmake@4.4.1%gcc@11.1.0~guile build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  6keg57t      ^mpfr@4.2.0%gcc@11.1.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  l3nt23x          ^autoconf-archive@2023.02.20%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  rmho7zt          ^texinfo@7.0%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  liqdych              ^gettext@0.21.1%gcc@11.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  kbnj5jn                  ^libxml2@2.10.3%gcc@11.1.0~python build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  qqx54or                  ^tar@1.34%gcc@11.1.0 build_system=autotools zip=pigz arch=linux-ubuntu20.04-x86_64
[+]  gojgmg4                      ^pigz@2.7%gcc@11.1.0 build_system=makefile arch=linux-ubuntu20.04-x86_64
[+]  v4yh5vm                      ^zstd@1.5.4%gcc@11.1.0~ipo+programs build_system=cmake build_type=RelWithDebInfo compression=none generator=make libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  vmymil3                  ^xz@5.4.1%gcc@11.1.0+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  bwi5kdp      ^openblas@0.3.21%gcc@11.1.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic~shared build_system=makefile patches=114f95f,d3d9b15 symbol_suffix=none threads=openmp arch=linux-ubuntu20.04-x86_64
[+]  muafle5          ^perl@5.36.0%gcc@11.1.0+cpanm+open+shared+threads build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  alkzfee              ^berkeley-db@18.1.40%gcc@11.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-x86_64
[+]  onsas2b              ^bzip2@1.0.8%gcc@11.1.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  qkfi67z              ^gdbm@1.23%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  lui3zbo                  ^readline@8.2%gcc@11.1.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64
[+]  hi2xsgc              ^zlib@1.2.13%gcc@11.1.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-x86_64
</pre></details>


Error:
```
$> spack -e . install
...
==> Installing suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa
==> No binary for suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/59/59c6ca2959623f0c69226cf9afb9a018d12a37fab3a8869db5f6d7f83b6b147d.tar.gz
==> No patches needed for suite-sparse
==> suite-sparse: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' '-C' 'UMFPACK' 'library' 'AUTOCC=no' 'CUDA=no' 'CUDA_PATH=' 'CFOPENMP=-fopenmp' 'CFLAGS=-O3 -fPIC' 'FFLAGS=-fPIC' 'F77FLAGS=-fPIC' 'MY_METIS_LIB=-L/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/metis-5.1.0-h32qvnmyrtvipy2jofpga4x6wwm4ml7g/lib -lmetis' 'MY_METIS_INC=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/metis-5.1.0-h32qvnmyrtvipy2jofpga4x6wwm4ml7g/include' 'BLAS=-L/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/openblas-0.3.21-bwi5kdpdcgmd4gnllykft6zuded25tgz/lib -lopenblas' 'LAPACK=-L/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/openblas-0.3.21-bwi5kdpdcgmd4gnllykft6zuded25tgz/lib -lopenblas' 'CFLAGS+=-fno-common -fexceptions' 'CMAKE_OPTIONS=-DCMAKE_INSTALL_PREFIX=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa -DCMAKE_LIBRARY_PATH=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa/lib' 'INSTALL=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa'

159 errors found in build log:
     1909    a - umfpack_zl_load_numeric.o
     1910    a - umfpack_zl_save_numeric.o
     1911    a - umfpack_zl_load_symbolic.o
     1912    a - umfpack_zl_save_symbolic.o
     1913    ranlib libumfpack.a
     1914    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/openblas-0.3.21-bwi5kdpdcgmd4gnllykft6zuded25tgz/lib/libopenblas.a(memory.o): in function `get_num_procs':
  >> 1915    memory.c:(.text+0x1f3): undefined reference to `omp_get_num_places'
     1916    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/openblas-0.3.21-bwi5kdpdcgmd4gnllykft6zuded25tgz/lib/libopenblas.a(dgemv.o): in function `dgemv_':
  >> 1917    gemv.c:(.text+0x390): undefined reference to `omp_get_max_threads'
  >> 1918    /usr/bin/ld: gemv.c:(.text+0x3bb): undefined reference to `omp_in_parallel'
     1919    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/openblas-0.3.21-bwi5kdpdcgmd4gnllykft6zuded25tgz/lib/libopenblas.a(dger.o): in function `dger_':
  >> 1920    ger.c:(.text+0x2aa): undefined reference to `omp_get_max_threads'
  >> 1921    /usr/bin/ld: ger.c:(.text+0x2d4): undefined reference to `omp_in_parallel'
     1922    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/openblas-0.3.21-bwi5kdpdcgmd4gnllykft6zuded25tgz/lib/libopenblas.a(dgemm.o): in function `dgemm_':
  >> 1923    gemm.c:(.text+0x491): undefined reference to `omp_get_max_threads'
  >> 1924    /usr/bin/ld: gemm.c:(.text+0x4a2): undefined reference to `omp_in_parallel'
...
  >> 2141    /usr/bin/ld: sasum.c:(.text+0x2b0): undefined reference to `omp_in_parallel'
     2142    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/openblas-0.3.21-bwi5kdpdcgmd4gnllykft6zuded25tgz/lib/libopenblas.a(srot_k_COOPERLAKE.o): in function `srot_k_COOPERLAKE':
  >> 2143    srot.c:(.text+0x332): undefined reference to `omp_get_max_threads'
  >> 2144    /usr/bin/ld: srot.c:(.text+0x350): undefined reference to `omp_in_parallel'
     2145    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/openblas-0.3.21-bwi5kdpdcgmd4gnllykft6zuded25tgz/lib/libopenblas.a(sgemm_direct_performant_COOPERLAKE.o): in function `sgemm_direct_performant_COOPERLAKE':
  >> 2146    sgemm_direct_performant.c:(.text+0x3f): undefined reference to `omp_get_max_threads'
  >> 2147    /usr/bin/ld: sgemm_direct_performant.c:(.text+0x4b): undefined reference to `omp_in_parallel'
  >> 2148    collect2: error: ld returned 1 exit status
  >> 2149    make[2]: *** [Makefile:297: /tmp/root/spack-stage/spack-stage-suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa/spack-src/lib/libumfpack.so.5.7.9] Error 1
     2150    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa/spack-src/UMFPACK/Lib'
  >> 2151    make[1]: *** [Makefile:19: library] Error 2
     2152    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa/spack-src/UMFPACK/Lib'
  >> 2153    make: *** [Makefile:19: library] Error 2
     2154    make: Leaving directory '/tmp/root/spack-stage/spack-stage-suite-sparse-5.13.0-6ptvibp3akmd7xzgswmg27jk47nwspxa/spack-src/UMFPACK'
```


### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.20.0.dev0 (b8735fd1e49006f5b48d516309168366298fe687)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/11099950/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/11099951/spack-build-out.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/suite-sparse/package.py
+++ b/var/spack/repos/builtin/packages/suite-sparse/package.py
@@ -95,6 +95,12 @@ class SuiteSparse(Package):
         "%gcc@:4.8", when="@5.2.0:", msg="gcc version must be at least 4.9 for suite-sparse@5.2.0:"
     )
 
+    def flag_handler(self, name, flags):
+        if name in ("cflags", "cxxflags"):
+            if self.spec.satisfies("^openblas ~shared threads=openmp"):
+                flags.append("-fopenmp")
+        return (flags, None, None)
+
     def symbol_suffix_blas(self, spec, args):
         """When using BLAS with a special symbol suffix we use defines to
         replace blas symbols, e.g. dgemm_ becomes dgemm_64_ when

--- a/var/spack/repos/builtin/packages/suite-sparse/package.py
+++ b/var/spack/repos/builtin/packages/suite-sparse/package.py
@@ -98,7 +98,7 @@ class SuiteSparse(Package):
     def flag_handler(self, name, flags):
         if name in ("cflags", "cxxflags"):
             if self.spec.satisfies("^openblas ~shared threads=openmp"):
-                flags.append("-fopenmp")
+                flags.append(self.compiler.openmp_flag)
         return (flags, None, None)
 
     def symbol_suffix_blas(self, spec, args):

